--- wayland-1.19.0/src/meson.build	2021-01-27 22:18:01.000000000 +0530
+++ wayland-1.19.0-fix/src/meson.build	2021-08-11 13:12:56.972776532 +0530
@@ -34,15 +34,15 @@
 
 	wayland_scanner_sources = [ 'scanner.c', 'dtddata.S' ]
 	wayland_scanner_includes = [ root_inc, protocol_inc ]
-
+	
 	wayland_scanner = executable(
-		'wayland-scanner',
-		wayland_scanner_sources,
-		c_args: [ '-include', 'config.h' ],
-		include_directories: wayland_scanner_includes,
-		dependencies: [ scanner_deps, wayland_util_dep, ],
-		install: true
-	)
+                        'wayland-scanner',
+                        wayland_scanner_sources,
+                        c_args: [ '-include', 'config.h' ],
+                        include_directories: wayland_scanner_includes,
+                        dependencies: [ scanner_deps, wayland_util_dep, ],
+                        install: true
+                        )
 
 	pkgconfig.generate(
 		name: 'Wayland Scanner',
@@ -58,13 +58,33 @@
 	)
 endif
 
-if meson.is_cross_build() or not get_option('scanner')
-	scanner_dep = dependency('wayland-scanner', native: true, version: meson.project_version())
-	wayland_scanner_for_build = find_program(scanner_dep.get_pkgconfig_variable('wayland_scanner'))
-else
-	wayland_scanner_for_build = wayland_scanner
+if meson.is_cross_build() and get_option('scanner')
+
+	scanner_dep_natve = dependency('expat, xml2', native: true, version: meson.project_version())
+	
+	wayland_util_native = static_library(
+        'wayland-util',
+        sources: 'wayland-util.c'
+	)
+
+	wayland_util_dep_native = declare_dependency(
+        link_with: wayland_util_native,
+        include_directories: include_directories('.')
+	)
+	
+	wayland_scanner = executable(
+                        'wayland-scanner-build',
+                        wayland_scanner_sources,
+                        c_args: [ '-include', 'config.h' ],
+                        include_directories: wayland_scanner_includes,
+                        dependencies: [ scanner_deps_native, wayland_util_dep_native, ],
+                        install: false,
+                        native : true
+                        )
 endif
 
+wayland_scanner_for_build = wayland_scanner
+
 if get_option('libraries')
 	# wayland libraries
 
